src/Spice86.Core/Emulator/InterruptHandlers/Dos/Xms/XmsBlock.cs

```diff
@@ -0,0 +1,100 @@
+﻿namespace Spice86.Core.Emulator.InterruptHandlers.Dos.Xms;
+public readonly struct XmsBlock : IEquatable<XmsBlock> {
+    public XmsBlock(int handle, uint offset, uint length, bool free) {
+        Handle = handle;
+        Offset = offset;
+        Length = length;
+        IsFree = free;
+    }
+    /// <summary>
+    /// Gets the handle which owns the block.
+    /// </summary>
+    public int Handle { get; }
+
+    /// <summary>
+    /// Gets the offset of the block from the XMS base address.
+    /// </summary>
+    public uint Offset { get; }
+    /// <summary>
+    /// Gets the length of the block in bytes.
+    /// </summary>
+    public uint Length { get; }
+
+    /// <summary>
+    /// Gets a value indicating whether the block is ready to be used.
+    /// </summary>
+    public bool IsFree { get; }
+
+
+    public override string ToString() {
+        if (!IsFree) {
+            return $"{Handle:X4}: {Offset:X8} to {Offset + Length:X8}";
+        } else {
+            return "Free";
+        }
+    }
+
+    public override bool Equals(object? obj) => obj is XmsBlock b && Equals(b);
+
+    public override int GetHashCode() => Handle ^ (int)Offset ^ (int)Length;
+
+    public bool Equals(XmsBlock other) => Handle == other.Handle && Offset == other.Offset && Length == other.Length && !IsFree == !other.IsFree;
+
+    /// <summary>
+    /// Allocates a block of memory from a free block.
+    /// </summary>
+    /// <param name="handle">Handle making the allocation.</param>
+    /// <param name="length">Length of the requested block in bytes.</param>
+    /// <returns>Array of blocks to replace this block.</returns>
+    public XmsBlock[] Allocate(int handle, uint length) {
+        if (!IsFree) {
+            throw new InvalidOperationException();
```

Comment:
A description ("Tried to allocate to an XMS block that is not free" or such) would be nice